File: api/validate.js
import { parseSpec } from '../core/turing_machine.js';

export default function handler(req, res) {
  try {
    const { spec } = req.body;

    if (!spec) {
      return res.status(400).json({
        success: false,
        error: 'Especificação vazia'
      });
    }

    const { tm, error } = parseSpec(spec);

    if (error) {
      return res.status(200).json({
        success: false,
        error
      });
    }

    return res.status(200).json({
      success: true,
      message: 'Especificação válida!',
      states: Array.from(tm.states),
      transitions_count: Object.keys(tm.transitions).length
    });
  } catch (e) {
    return res.status(500).json({
      success: false,
      error: `Erro no servidor: ${e.message}`
    });
  }
}

File: api/validate.py
from core.turing_machine import parse_spec
import json
import sys
import os

# Adiciona o diretório raiz ao path para importar o módulo core
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))


def handler(event, context):
    """Valida a especificação da máquina de Turing"""
    try:
        # Parse do body da requisição
        if isinstance(event.get('body'), str):
            body = json.loads(event['body'])
        else:
            body = event.get('body', {})

        spec = body.get('spec', '')

        if not spec:
            return {
                'statusCode': 400,
                'headers': {
                    'Content-Type': 'application/json',
                    'Access-Control-Allow-Origin': '*'
                },
                'body': json.dumps({
                    'success': False,
                    'error': 'Especificação vazia'
                })
            }

        tm, error = parse_spec(spec)

        if error:
            return {
                'statusCode': 200,
                'headers': {
                    'Content-Type': 'application/json',
                    'Access-Control-Allow-Origin': '*'
                },
                'body': json.dumps({
                    'success': False,
                    'error': error
                })
            }

        return {
            'statusCode': 200,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*'
            },
            'body': json.dumps({
                'success': True,
                'message': 'Especificação válida!',
                'states': list(tm.states),
                'transitions_count': len(tm.transitions)
            })
        }

    except Exception as e:
        return {
            'statusCode': 500,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*'
            },
            'body': json.dumps({
                'success': False,
                'error': f'Erro no servidor: {str(e)}'
            })
        }
